### DISCUSSION TITLE
Correct way to identify a "fetch error"

### DISCUSSION THREAD

**Discussion by janeklb** [nodejs/help#4561] [Q&A]
_using Node.js 18.17.1_ 

After switching from `node-fetch` to the native fetch now available in node, im unsure how to best identify "fetch errors".

In `node-fetch` there was such a thing as [`FetchError`](https://github.com/node-fetch/node-fetch/blob/main/README.md#class-fetcherror). This is of course not a whatwg fetch spec concept -- just something that `node-fetch` included as an extension. That said, it was handy because it allowed me to easily determine if something was a "fetch error" by simply evaluating `error instanceof FetchError` (upon which i could access `error.code` etc. while keeping TypeScript happy).

Using the native `fetch`, errors appear as follows:

```
> fetch('http://xxx.yyy.com').catch((e) => console.log(e));
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 1867,
  [Symbol(trigger_async_id_symbol)]: 1840
}
> TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: Error: getaddrinfo ENOTFOUND xxx.yyy.com
      at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:108:26)
      at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -3008,
    code: 'ENOTFOUND',
    syscall: 'getaddrinfo',
    hostname: 'xxx.yyy.com'
  }
}
```

Could anyone help with the following Qs:
1/ What's the reasoning behind this being a `TypeError`? Feels misleading..
2/ Is the best way to identify failed fetches to check `error instanceof TypeError` and `error.message === 'failed fetch'`, and then retrieve `error.cause`?
3/ Is there a (internal?) type for `error.cause`? I'm trying to come up with a good (TypeScript-safe) way to get at the `error.type` and `error.code` properties in order to tag my application metrics, and am not sure if there's an existing type that fits the cause (I'm aware of the [megathread](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924) in DT about having TS types for `fetch` but wondering if there's maybe an undici type that is exposed somewhere?)

**Comment by axel7083** [comment]
I also face similar issue, when working with https://github.com/octokit/request-error.js they seems to pass fetch underlining error to the cause, so we get something like

```
> error.cause.cause
Error: getaddrinfo EAI_AGAIN api.github.com
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:120:26)
    at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
  errno: -3001,
  code: 'EAI_AGAIN',
  syscall: 'getaddrinfo',
  hostname: 'api.github.com'
```

I was looking for a way to type the fetch error, so I can check for which syscall failed, and _guess_ that the problem is related to network issue.

**Comment by k4hvecii** [comment]
The TypeError is intentional here because Node's fetch aims to follow the Web Fetch API rather than exposing a Node-specific FetchError class.

I would not rely on:

error.message === "fetch failed"

because the message is an implementation detail and is unnecessarily brittle.

For application-level handling, I would first treat the rejected value as an unknown Error, and then inspect cause only when I need Node/Undici-specific diagnostics.

For example:

try {
  await fetch(url);
} catch (error) {
  if (error instanceof TypeError) {
    const cause = error.cause;

    if (
      cause &&
      typeof cause === "object" &&
      "code" in cause
    ) {
      console.log("network error code:", cause.code);
    }
  }
}

So there are really two layers here:

1. The Fetch API surface exposes a TypeError for a failed fetch.
2. Node's Undici implementation may attach the underlying networking error as error.cause, for example ENOTFOUND, ECONNREFUSED, UND_ERR_SOCKET, etc.

I would therefore avoid treating TypeError + the exact "fetch failed" message as a stable FetchError equivalent.

If my application needed strongly typed, Node-specific network error codes as part of its public logic rather than only diagnostics/metrics, I would probably use Undici directly (or wrap fetch behind my own application error type) instead of depending on the internal shape of fetch's cause.

For TypeScript, a small type guard around cause is probably safer than trying to import an internal Node type, since the exact underlying cause is implementation-specific.
